### PR TITLE
feat(meta): support BTreeMapTransaction that avoids copying the whole BTreeMap

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Currently in our hummock manager, we use `VarTransaction<BTreeMap>` to wrap a hummock manager state that is of BTreeMap type to do some transactional work. The current implementation of `VarTransaction<BTreeMap>` will take a copy of the BTreeMap and all modifications will be applied to the BTreeMap copy. When it is applied to a transaction, we will iterate the BTreeMap copy and the original BTreeMap and apply the upsert and delete by comparing the two values of the same key one by one. Upon commit the BTreeMap copy will replace the original BTreeMap. Although we have `BTreeMapEntryTransaction` to avoid copying the BTreeMap but still able to update the BTreeMap, it only allows modifying one BTreeMap entry. In case we want to insert or delete multiple entries in a transaction, we still need the `VarTransaction<BTreeMap>`.

However, taking a copy of BTreeMap may impact the performance when the `BTreeMap` gets large. In a previous commit 4a75d24972639f00d2ad4f36d298d4075ceb7e7c on the main branch, a unit test named `hummock::compactor_tests::tests::test_compaction_drop_key_by_ttl` commit 1000 epochs in total (the unit test has be modified recently to committing much less than 1000 times and it is much faster), with 1 SST added to HummockVersion in each epoch. All operations are in-memory, and it should not take too much time. However, the unit test takes about 5 seconds to finish. The flamegraph shows that half of the time is spent on BTreeMap related stuffs (the rest half are protobuf encoding). 
image

In this PR, we introduce a `BTreeMapTransaction`, which supports basic BTreeMap operation such as `get`, `get_mut`, `insert` and `remove`. It has a field `staging` to record all the `insert`, `remove` and `update`(generated by modification on `get_mut`) operations on the BTreeMap entry. When we read the BTreeMap with `get` and `get_mut`, it will merge the value in `staging` and the original tree. Upon commit it will apply the change on `staging` to the original BTreeMap. With this PR, the unit test mentioned takes 2 seconds to finish, and in the flamegraph we can barely see any BTreeMap operation.
image

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

**Comment by Little-Wallace** [comment]
Would you mind give more information such as CPU of meta-service ?

File: src/meta/src/model/mod.rs [hunk 1]
```diff
@@ -236,34 +238,207 @@ where
...
+    Insert(V),
+    Delete,
+}
+
+/// A `ValTransaction` that wraps a `BTreeMap`. It supports basic `BTreeMap` operations like `get`,
+/// `get_mut`, `insert` and `remove`. Incremental modification of `insert`, `remove` and `get_mut`
+/// are stored in `staging`. On `commit`, it will apply the changes stored in `staging` to the in
+/// memory btree map. When serve `get` and `get_mut`, it merges the value stored in `staging` and
+/// `tree_ref`.
+pub struct BTreeMapTransaction<'a, K: Ord, V> {
```
Comment (hzxa21) [inline src/meta/src/model/mod.rs:319]: A big portion of the `BTreeMapTransaction` and `BTreeMapTransactionValueGuard` implementation is to deal with merging "staging" entry with "original" entry. Do we actually have a use case that we need to iter/read the BTreeMap after writing "staging" entry to it?

**Review by hzxa21 — APPROVED** [review]
Discussed offline. It is another medium-size refactoring to remove `get_mut` if we want to get rid of the merge logic in BTreeMapTransaction. We think it is not worthwhile so the plan is to get this PR merged first.

**Comment by wenym1** [comment]
Provide some comparison metrics:
Current main:
image
image

With this PR
image
image

With this PR, the memory usage meta decreases compared with the current main branch, though the memory usage is low. And since the bottleneck for real process time is committing ETCD, the overall real process time is basically the same in both cases.

Since the benchmark is not long-running, the state stored in meta is relatively small. But I think while the system gets long-running and the state becomes large, we can benefit from avoiding copying the whole BTreeMap.

**Pull request merged**